account_invoice_margin/models/account_invoice.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# © 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, fields, models
+import openerp.addons.decimal_precision as dp
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    margin = fields.Float(
+        compute='_compute_margin',
+        digits=dp.get_precision('Product Price'),
+        store=True,
+        string='Margin',
+    )
+    margin_percent = fields.Float(
+        string='% Margin',
+        compute='_compute_margin',
+        store=True,
+        readonly=True,
+    )
+    purchase_price = fields.Float(
+        digits=dp.get_precision('Product Price'),
+        string='Cost',
+    )
+
+    @api.multi
+    @api.depends('purchase_price', 'price_subtotal')
+    def _compute_margin(self):
+        for line in self.filtered(lambda x: x.invoice_id.type[:2] != 'in'):
+            tmp_margin = line.price_subtotal - (
+                line.purchase_price * line.quantity)
+            if line.invoice_id.type == 'out_refund':
+                tmp_margin *= -1
+            line.update({
+                'margin': tmp_margin,
+                'margin_percent': (tmp_margin / line.price_subtotal * 100.0 if
+                                   line.price_subtotal else 0.0),
+            })
+
+    @api.onchange('product_id', 'uom_id')
+    def _onchange_product_id_margin(self):
+        if self.invoice_id.type in ['out_invoice', 'out_refund']:
+            purchase_price = self.product_id.standard_price
```

Comment:
This doesn't take into account if the price is with taxes included. Put it on known issues or change the calculation.

Reply:
Yeah, but it doesn't convert that in the correct behavior.

Reply:
What about a field standard_price_vat_excl ? In a new module, and make account invoice margin and pos margin, depending on the new module ?

But are you sure that its possible to have a standard price vat incl ? I never seen any implementation with such settings. Could you explain when it can occur ?

Regards.

Reply:
You can easily see it:

* Define a purchase tax vat included.
* Put it in a product.
* Define cost price (i.e. 100).
* Include that product on a purchase order

You will see that the standard price, it's considered as vat included, so the real one to use for margins is 86.96:

![seleccion_009](https://user-images.githubusercontent.com/7165771/29718035-aa21eea8-89b1-11e7-86f6-cfa3c67bb7c3.png)

The other way around can't happen as method `compute_taxes` already excludes taxes.

Reply:
OK ! Thanks. (I never used purchase tax vat included, but if it's a use case, it should be handled.

So what about a standard_price_vat_excl field, as this is what we need here ?

If agree, I can make a PR to do that.

Reply:
I think we don't need an extra module. Just call <purchase_tax> compute taxes method with the cost price and the quantity and use the untaxed amount returned.